File: cpp/src/thread_pool.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <functional>
#include <chrono>
#include <iomanip>
#include <atomic>

const size_t NUM_TASKS = 100'000;
const size_t NUM_WORKERS = 8;

class ThreadPool {
public:
    ThreadPool(size_t num_threads) : stop(false) {
        for (size_t i = 0; i < num_threads; i++) {
            workers.emplace_back([this] {
                while (true) {
                    std::function<void()> task;
                    {
                        std::unique_lock<std::mutex> lock(queue_mutex);
                        condition.wait(lock, [this] { return stop || !tasks.empty(); });
                        
                        if (stop && tasks.empty()) {
                            return;
                        }
                        
                        task = std::move(tasks.front());
                        tasks.pop();
                    }
                    task();
                }
            });
        }
    }
    
    template<class F>
    void enqueue(F&& f) {
        {
            std::unique_lock<std::mutex> lock(queue_mutex);
            tasks.emplace(std::forward<F>(f));
        }
        condition.notify_one();
    }
    
    ~ThreadPool() {
        {
            std::unique_lock<std::mutex> lock(queue_mutex);
            stop = true;
        }
        condition.notify_all();
        for (std::thread& worker : workers) {
            worker.join();
        }
    }
    
private:
    std::vector<std::thread> workers;
    std::queue<std::function<void()>> tasks;
    std::mutex queue_mutex;
    std::condition_variable condition;
    bool stop;
};

uint64_t heavy_computation(size_t n) {
    uint64_t result = 0;
    for (size_t i = 0; i < 1000; i++) {
        result += static_cast<uint64_t>(n) * i;
    }
    return result;
}

int main() {
    std::atomic<uint64_t> counter(0);
    
    // Warm-up
    {
        ThreadPool pool(NUM_WORKERS);
        std::atomic<uint64_t> counter_warmup(0);
        for (size_t i = 0; i < 100; i++) {
            pool.enqueue([i, &counter_warmup] {
                uint64_t result = heavy_computation(i);
                counter_warmup += result;
            });
        }
    }
    
    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    {
        ThreadPool pool(NUM_WORKERS);
        
        for (size_t i = 0; i < NUM_TASKS; i++) {
            pool.enqueue([i, &counter] {
                uint64_t result = heavy_computation(i);
                counter += result;
            });
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    
    std::chrono::duration<double> duration = end - start;
    
    std::cout << std::fixed << std::setprecision(6) << duration.count() << std::endl;
    std::cerr << "Final count: " << counter.load() << std::endl;
    
    return 0;
}

File: rust/src/thread_pool.rs
use std::sync::{Arc, Mutex, mpsc};
use std::thread;
use std::time::Instant;

const NUM_TASKS: usize = 100_000;
const NUM_WORKERS: usize = 8;

struct ThreadPool {
    workers: Vec<thread::JoinHandle<()>>,
    sender: Option<mpsc::Sender<Task>>,
}

type Task = Box<dyn FnOnce() + Send + 'static>;

impl ThreadPool {
    fn new(size: usize) -> ThreadPool {
        let (sender, receiver) = mpsc::channel::<Task>();
        let receiver = Arc::new(Mutex::new(receiver));
        let mut workers = Vec::with_capacity(size);
        
        for _ in 0..size {
            let receiver = Arc::clone(&receiver);
            let handle = thread::spawn(move || {
                loop {
                    let task = receiver.lock().unwrap().recv();
                    match task {
                        Ok(task) => task(),
                        Err(_) => break,
                    }
                }
            });
            workers.push(handle);
        }
        
        ThreadPool { workers, sender: Some(sender) }
    }
    
    fn execute<F>(&self, f: F)
    where
        F: FnOnce() + Send + 'static,
    {
        self.sender.as_ref().unwrap().send(Box::new(f)).unwrap();
    }
}

impl Drop for ThreadPool {
    fn drop(&mut self) {
        // Drop the sender to close the channel
        // This will cause recv() to return Err and workers to exit
        drop(self.sender.take());
        
        // Now join all workers
        while let Some(worker) = self.workers.pop() {
            worker.join().unwrap();
        }
    }
}

fn heavy_computation(n: usize) -> u64 {
    let mut result = 0u64;
    for i in 0..1000 {
        result = result.wrapping_add((n as u64).wrapping_mul(i));
    }
    result
}

fn main() {
    let counter = Arc::new(Mutex::new(0u64));
    
    // Warm-up
    {
        let pool = ThreadPool::new(NUM_WORKERS);
        let counter_warmup = Arc::new(Mutex::new(0u64));
        for i in 0..100 {
            let counter_clone = Arc::clone(&counter_warmup);
            pool.execute(move || {
                let result = heavy_computation(i);
                let mut count = counter_clone.lock().unwrap();
                *count = count.wrapping_add(result);
            });
        }
    }
    
    // Benchmark
    let start = Instant::now();
    {
        let pool = ThreadPool::new(NUM_WORKERS);
        
        for i in 0..NUM_TASKS {
            let counter_clone = Arc::clone(&counter);
            pool.execute(move || {
                let result = heavy_computation(i);
                let mut count = counter_clone.lock().unwrap();
                *count = count.wrapping_add(result);
            });
        }
    }
    let duration = start.elapsed();
    
    let final_count = *counter.lock().unwrap();
    
    println!("{:.6}", duration.as_secs_f64());
    eprintln!("Final count: {}", final_count);
}

